clickhouse-admin/src/context.rs

```diff
@@ -3,84 +3,555 @@
     clickward: Clickward,
-    oximeter_client: OximeterClient,
-    initialization_lock: Arc<Mutex<()>>,
+    clickhouse_address: SocketAddrV6,
     log: Logger,
+    pub generate_config_tx: Sender<GenerateConfigRequest>,
+    pub generation_rx: watch::Receiver<Option<Generation>>,
+    pub db_init_tx: Sender<DbInitRequest>,
 }
 
 impl ServerContext {
-    pub fn new(clickhouse_cli: ClickhouseCli) -> Self {
-        let ip = clickhouse_cli.listen_address.ip();
-        let address = SocketAddrV6::new(*ip, CLICKHOUSE_TCP_PORT, 0, 0);
-        let oximeter_client =
-            OximeterClient::new(address.into(), &clickhouse_cli.log);
+    pub fn new(
+        log: &Logger,
+        binary_path: Utf8PathBuf,
+        listen_address: SocketAddrV6,
+    ) -> Result<Self> {
+        let clickhouse_cli =
+            ClickhouseCli::new(binary_path, listen_address, log);
+
+        let ip = listen_address.ip();
+        let clickhouse_address =
+            SocketAddrV6::new(*ip, CLICKHOUSE_TCP_PORT, 0, 0);
         let clickward = Clickward::new();
-        let log =
-            clickhouse_cli.log.new(slog::o!("component" => "ServerContext"));
-        Self {
+        let log = log.new(slog::o!("component" => "ServerContext"));
+
+        let config_path = Utf8PathBuf::from_str(CLICKHOUSE_SERVER_CONFIG_DIR)
+            .unwrap()
+            .join(CLICKHOUSE_SERVER_CONFIG_FILE);
+
+        // If there is already a configuration file with a generation number we'll
+        // use that. Otherwise, we set the generation number to None.
+        let gen = read_generation_from_file(config_path)?;
+        let (generation_tx, generation_rx) = watch::channel(gen);
+
+        // We only want to handle one in flight request at a time. Reconfigurator execution will retry
+        // again later anyway. We use flume bounded channels with a size of 0 to act as a rendezvous channel.
+        let (generate_config_tx, generate_config_rx) = flume::bounded(0);
+        tokio::spawn(long_running_generate_config_task(
+            generate_config_rx,
+            generation_tx,
+        ));
+
+        let (db_init_tx, db_init_rx) = flume::bounded(0);
+        tokio::spawn(long_running_db_init_task(db_init_rx));
+
+        Ok(Self {
             clickhouse_cli,
             clickward,
-            oximeter_client,
-            initialization_lock: Arc::new(Mutex::new(())),
+            clickhouse_address,
             log,
-        }
+            generate_config_tx,
+            generation_rx,
+            db_init_tx,
+        })
+    }
+
+    pub async fn send_generate_config_and_enable_svc(
+        &self,
+        replica_settings: ServerConfigurableSettings,
+    ) -> Result<GenerateConfigResult, HttpError> {
+        let clickward = self.clickward();
+        let log = self.log();
+        let node_settings = NodeSettings::Replica {
+            settings: replica_settings,
+            fmri: "svc:/oxide/clickhouse_server:default".to_string(),
```

Comment:
This method looks basically identical to the one above on `KeeperServerContext`, except for the `NodeSettings` here. Since most of the method is error handling, do you think it would make sense to extract the common bits into a free-standing function? Then both methods could be something like

```rust
let node_settings = /* ... the right node settings for this context ... */;
generate_config_and_enable_svc(node_settings, self.clickward(), self.log()).await
```